src/vs/workbench/services/remote/common/remoteExplorerService.ts

```diff
@@ -10,15 +10,120 @@ import { registerSingleton } from 'vs/platform/instantiation/common/extensions';
+import { Disposable } from 'vs/base/common/lifecycle';
 
 export const IRemoteExplorerService = createDecorator<IRemoteExplorerService>('remoteExplorerService');
 export const REMOTE_EXPLORER_TYPE_KEY: string = 'remote.explorerType';
 
+export interface Tunnel {
+	remote: number;
+	localUri: URI;
+	local?: number;
+	name?: string;
+	description?: string;
+	closeable?: boolean;
+}
+
+export class TunnelModel extends Disposable {
+	forwarded: Map<number, Tunnel>;
+	published: Map<number, Tunnel>;
+	candidates: Map<number, Tunnel>;
+	private _onForwardPort: Emitter<Tunnel> = new Emitter();
+	public onForwardPort: Event<Tunnel> = this._onForwardPort.event;
+	private _onClosePort: Emitter<number> = new Emitter();
+	public onClosePort: Event<number> = this._onClosePort.event;
+	private _onPortName: Emitter<number> = new Emitter();
+	public onPortName: Event<number> = this._onPortName.event;
+	constructor(
+		@ITunnelService private readonly tunnelService: ITunnelService
+	) {
+		super();
+		this.forwarded = new Map();
+		this.tunnelService.tunnels.then(tunnels => {
+			tunnels.forEach(tunnel => {
+				if (tunnel.localAddress) {
+					this.forwarded.set(tunnel.tunnelRemotePort, {
+						remote: tunnel.tunnelRemotePort,
+						localUri: tunnel.localAddress,
+						local: tunnel.tunnelLocalPort
+					});
+				}
+			});
+		});
+
+		this.published = new Map();
+		this.candidates = new Map();
+		this._register(this.tunnelService.onTunnelOpened(tunnel => {
+			if (this.candidates.has(tunnel.tunnelRemotePort)) {
+				this.candidates.delete(tunnel.tunnelRemotePort);
+			}
+			if (!this.forwarded.has(tunnel.tunnelRemotePort) && tunnel.localAddress) {
+				this.forwarded.set(tunnel.tunnelRemotePort, {
+					remote: tunnel.tunnelRemotePort,
+					localUri: tunnel.localAddress,
+					local: tunnel.tunnelLocalPort
+				});
+			}
+			this._onForwardPort.fire(this.forwarded.get(tunnel.tunnelRemotePort)!);
+		}));
+		this._register(this.tunnelService.onTunnelClosed(remotePort => {
+			if (this.forwarded.has(remotePort)) {
+				this.forwarded.delete(remotePort);
+				this._onClosePort.fire(remotePort);
+			}
+		}));
+	}
+
+	async forward(remote: number, local?: number, name?: string): Promise<void> {
+		if (!this.forwarded.has(remote)) {
+			const tunnel = await this.tunnelService.openTunnel(remote, local);
+			if (tunnel && tunnel.localAddress) {
+				const newForward: Tunnel = {
+					remote: tunnel.tunnelRemotePort,
+					local: tunnel.tunnelLocalPort,
+					name: name,
+					closeable: true,
+					localUri: tunnel.localAddress
+				};
+				this.forwarded.set(remote, newForward);
+				this._onForwardPort.fire(newForward);
+			}
+		}
+	}
+
+	name(remote: number, name: string) {
+		if (this.forwarded.has(remote)) {
+			this.forwarded.get(remote)!.name = name;
+			this._onPortName.fire(remote);
+		}
+	}
+
+	async close(remote: number): Promise<void> {
+		return this.tunnelService.closeTunnel(remote);
+	}
+
+	address(remote: number): URI | undefined {
```

Comment:
[nit] This can be simplified to some form of ` (this.forwarded.get(remote) || this.published.get(remote))?.localUri`